initfiles/bin/init_thor

```diff
@@ -15,26 +15,197 @@
+INSTALL_DIR=$(dirname ${PATH_PRE})/..
+source  ${INSTALL_DIR}/etc/init.d/hpcc_common
+
+PID_NAME="$PID/$(basename $PWD).pid"
+INIT_PID_NAME="$PID/init_$(basename $PWD).pid"
 echo $$ > $INIT_PID_NAME
 
 export SENTINEL="thor.sentinel"
 rm -f ${SENTINEL}
 
-killed() {
-        echo "Stopping"
-        kill_process ${SENTINEL} ${PID_NAME} 3
-        exit 255
+if [[ -z "$deploydir" ]]; then
+    deploydir=$(pwd -P)
+fi
+
+instancedir=$(pwd -P)
+source $instancedir/setvars
+
+if [[ ! -z ${THORPRIMARY} ]]; then
+    groupName=${THORPRIMARY}
+else
+    groupName=${THORNAME}
+fi
+ln -s -f $deploydir/thormaster${LCR} thormaster_$THORNAME
+
+ENV_DIR=$(cat ${HPCC_CONFIG} | sed -n "/\[DEFAULT\]/,/\[/p" | grep "^configs=" | sed -e 's/^configs=//')
+logdir=$(updtdalienv $ENV_DIR/environment.xml -d log thor $THORNAME)
+logfile=$logdir/${THORNAME}.log
+
+contains()
+{
+    local n=$#
+    local value=${!n}
+    local i=0
+    for ((i=1;i < ${n};i++)); do
+        if [[ "${!i}" == "${value}" ]]; then
+            echo "${i}"
+            return 0
+        fi
+    done
+    echo "0"
+    return 1
+}
+
+makethorgroup()
+{
+    if [[ -z "$localthorportinc" ]]; then
+        localthorportinc=200
+    fi
+    rm -f thorgroup
+
+    declare -a ports_used
+    declare -a hosts
+    for slave in $(cat slaves); do
+        p=$(contains "${hosts[@]}" "${slave}")
+        if [[ 0 == ${p} ]]; then
+            echo "${slave}:${THORSLAVEPORT}" >> thorgroup
+            p=$((${#hosts[@]}+1))
+            ports[${p}]=${THORSLAVEPORT}
+            hosts=(${hosts[@]} $slave)
+        else
+            newport=$((${ports[${p}]}+${localthorportinc}))
+            echo "${slave}:${newport}" >> thorgroup
+            ports[${p}]=${newport}
+        fi
+    done
+}
+
+kill_slaves()
+{
+    if [[ "$localthor" = "true" ]]; then
+        SLAVE_PID_NAME="$PID/${comp_base}_slave_*.pid"
+        slavepid=$(cat ${SLAVE_PID_NAME} 2> /dev/null)
+        kill -9 $slavepid 2> /dev/null
+    else
+        # we want to kill only slaves that have already been started in run_thor
+        if [[ -r $instancedir/uslaves.start ]]; then
+            nslaves=$(cat $instancedir/uslaves.start 2> /dev/null | wc -l)
+            $deploydir/frunssh $instancedir/uslaves.start "/bin/sh -c '$deploydir/init_thorslave %a $THORMASTER $THORMASTERPORT $logdir $instancedir $deploydir $THORNAME $PATH_PRE $logredirect stop'" -i:$SSHidentityfile -u:$SSHusername -pe:$SSHpassword -t:$SSHtimeout -a:$SSHretries -n:$nslaves 2>&1 | egrep -v "no process killed"
+            echo slaves stopped
+        fi
+    fi
+
+    rm -f $instancedir/uslaves.start > /dev/null 2>&1
+}
+
+killed()
+{
+    echo "Stopping"
+    kill_slaves
+    kill_process ${PID_NAME} thormaster_${comp_base} 8
+    rm -f $INIT_PID_NAME > /dev/null 2>&1
+    exit 255
 }
 
-echo $$ > $PID_NAME
-$deploydir/start_thor $deploydir &
 trap "killed" SIGINT SIGTERM
-wait
-rm $PID_NAME
+# attempt to clean up any old slaves
+kill_slaves
```

Comment:
in startCmd we run cleanup_component if any unhealthy return value comes back from check_status.  This means we should never enter a new init script without already doing an attempted SIGTERM and SIGKILL on any group pid that is associated with that component and somehow still alive.   The reason that I think we still of course should do a sigkill on the slaves though is that since slaves aren't generic to all the other components, it isn't written into cleanup_component.  So there could somehow, someway, be a slave sitting on one of our machines that's just deadlocked.  So the kill_slaves that is done here is technically called after a stop() type of function is called, but from within hpcc_init, and it gets called for every component if check_status returns anything other than a 0 (healthy and running) or 1 (healthy stopped.)